### PR TITLE
add new FileSystemUtils class

### PR DESCRIPTION
**Pull request by ximex**
A class that handles nearly all file system stuff and has nearly all checks (is writable, ...)
This class should replace `MyFiles` and `Folder` classes.

### CODE REVIEW

File: adm_program/system/classes/filesystemutils.php [hunk 1]
```diff
@@ -0,0 +1,1016 @@
...
+    const ROOT_ID = 0;
+    const ROOT_FOLDER = '/';
+
+    private static $workingDirectories = array();
+
+    /**
+     * @param array<int,string> $directoryPaths
+     * @throws \UnexpectedValueException
+     */
+    public static function setWorkingDirectories(array $directoryPaths = array())
```
Comment (thomas-rcv) [inline adm_program/system/classes/filesystemutils.php]: Name the method like "setValidDirectories" Sounds better for me. "Working" is understood in other context

File: adm_program/system/classes/filesystemutils.php [hunk 2]
```diff
@@ -0,0 +1,1016 @@
...
+        }
+
+        self::$workingDirectories = $directoryPaths;
+    }
+
+    /**
+     * @param string $path
+     * @throws \RuntimeException
+     */
+    private static function checkIsInWorkingDirectories($path)
```
Comment (thomas-rcv) [inline adm_program/system/classes/filesystemutils.php]: Working -> Valid ?

File: adm_program/system/classes/filesystemutils.php [hunk 3]
```diff
@@ -0,0 +1,1016 @@
...
+
+        foreach (self::$workingDirectories as $workingDirectory)
+        {
+            if (strpos($path, $workingDirectory) === 0)
+            {
+                return;
+            }
+        }
+
+        throw new \RuntimeException('Path is not in working directory!');
```
Comment (thomas-rcv) [inline adm_program/system/classes/filesystemutils.php]: Path is not a valid directory

File: adm_program/system/classes/filesystemutils.php [hunk 4]
```diff
@@ -0,0 +1,1016 @@
...
+
+        if (!file_exists($path))
+        {
+            throw new \UnexpectedValueException('Path does not exist!');
+        }
+
+        $fileOwnerResult = fileowner($path);
+        if ($fileOwnerResult === false)
+        {
+            throw new \RuntimeException('File owner could not be determined!');
```
Comment (thomas-rcv) [inline adm_program/system/classes/filesystemutils.php]: Please Change "could not" to "cannot". It´s more proper speaking

File: adm_program/system/classes/filesystemutils.php [hunk 5]
```diff
@@ -0,0 +1,1016 @@
...
+
+        if (!file_exists($path))
+        {
+            throw new \UnexpectedValueException('Path does not exist!');
+        }
+
+        $fileGroupResult = filegroup($path);
+        if ($fileGroupResult === false)
+        {
+            throw new \RuntimeException('File group could not be determined!');
```
Comment (thomas-rcv) [inline adm_program/system/classes/filesystemutils.php]: could not -> cannot

File: adm_program/system/classes/filesystemutils.php [hunk 6]
```diff
@@ -0,0 +1,1016 @@
...
+
+        if (!file_exists($path))
+        {
+            throw new \UnexpectedValueException('Path does not exist!');
+        }
+
+        $perms = fileperms($path);
+        if ($perms === false)
+        {
+            throw new \RuntimeException('File permissions could not be read!');
```
Comment (thomas-rcv) [inline adm_program/system/classes/filesystemutils.php]: could not -> cannot

File: adm_program/system/classes/filesystemutils.php [hunk 7]
```diff
@@ -0,0 +1,1016 @@
...
+        }
+        if (!is_readable($directoryPath))
+        {
+            throw new \UnexpectedValueException('Directory is not readable!');
+        }
+
+        $dirHandle = opendir($directoryPath);
+        if ($dirHandle === false)
+        {
+            throw new \RuntimeException('Directory could not be opened!');
```
Comment (thomas-rcv) [inline adm_program/system/classes/filesystemutils.php]: could not -> cannot

File: adm_program/system/classes/filesystemutils.php [hunk 8]
```diff
@@ -0,0 +1,1016 @@
...
+        $flags = 0;
+        if ($append)
+        {
+            $flags = FILE_APPEND;
+        }
+
+        $bytes = file_put_contents($filePath, $data, $flags);
+        if ($bytes === false)
+        {
+            throw new \RuntimeException('File could not be written!');
```
Comment (thomas-rcv) [inline adm_program/system/classes/filesystemutils.php]: could not -> cannot

File: adm_program/system/classes/filesystemutils.php (hunk 1 above)
Comment (ximex) [reply adm_program/system/classes/filesystemutils.php]: Maybe "Allowed" instead of "Valid"?

Comment (thomas-rcv) [reply adm_program/system/classes/filesystemutils.php]: Yes, also ok

**Comment by ximex** [comment]
@thomas-rcv please review the changes
@Fasse any comments?

**Comment by thomas-rcv** [comment]
You can merge

**Comment by Fasse** [comment]
I will look at it this evening...

**Comment by Fasse** [comment]
> Some example of usage in comment head and more descriptions of the method are helpfull -> DokuWiki
This is not done yet. For me it's important that we have a well documented code. Therefore new classes should come with a good method description and also parameter and return description. Please add this to the FileSystemUtils class because now there is no documentation.

I will do today some functional tests.

**Comment by ximex** [comment]
@Fasse or @thomas-rcv please look at the phpdocs and the new normalizePath method. If all is ok please merge

**Comment by Fasse** [comment]
You now add a very short description to all method.
>     * Gets info about the process owner
>     * @return array<string,string|int> Returns info about the process owner
>    public static function getProcessOwnerInfo()

So I ask myself if I read this: Which prozess what owner ? Also I don't know how the return string looks like. What is the info?

Please write a detailed description so that people know what that function does and can use it. For me this class is a blackbox until I read the code.

File: adm_program/system/classes/filesystemutils.php [hunk 9]
```diff
@@ -0,0 +1,1154 @@
...
+    /**
+     * Get a normalized path
+     * @param string $path The path to normalize
+     * @return string The normalized path
+     */
+    public static function getNormalizedPath($path)
+    {
+        // Normalize directory separators
+        $path = str_replace('\\', '/', $path);
+        $path = preg_replace('/\/+/', '/', $path);
```
Comment (Fasse) [inline adm_program/system/classes/filesystemutils.php]: please describe what do you want to do with this.

File: adm_program/system/classes/filesystemutils.php [hunk 10]
```diff
@@ -0,0 +1,1154 @@
...
+    const ROOT_ID = 0;
+    const ROOT_FOLDER = '/';
+
+    /**
+     * @var array<int,string> The allowed directories
+     */
+    private static $allowedDirectories = array();
+
+    /**
+     * Get a normalized path
```
Comment (Fasse) [inline adm_program/system/classes/filesystemutils.php]: Please describe why to use this function and what is the advantage.

**Comment by ximex** [comment]
@Fasse enough improvements?

**Comment by Fasse** [comment]
@ximex yes, you can merge.

**Pull request merged**